File: pa1/Shuffle.java
//
//  Shuffle.java
//  
//  Created by Colin Rognlie on 6/29/05.
//  Copyright 2005. All rights reserved.
//
//  Colin Rognlie, reflex@cats.ucsc.edu, CMPS101 pa1
//

public class Shuffle {
	// Fields
	private List deck = new List();
	private List initdeck = new List();
	private int shuffles;

	// Constructor
	// public Shuffle (int n)
	// tests number of cards, initializes variables
	// shuffles until return to test state
	// prints number of required shuffles
	public Shuffle (int n) {
		if ( n < 1 ) { throw new RuntimeException("Argument is less than 1"); }
		for (int i = 1; i <= n; i++) { deck.insertAfterLast(i); }
		initdeck = deck.copy();
		shuffles = 0;
		// Shuffle until deck returns to initial state
		for( deckShuffle(deck); !deck.equals(initdeck); deckShuffle(deck) ) {}
		// Print number of shuffles
		System.out.println(shuffles);
	}

	// Manipulation Procedure
	// private void deckShuffle(List L)
	// shuffles the deck L according to the given algorithm, increments shuffles
	private void deckShuffle(List L) {
		List A = new List();
		List B = new List();
		// Deal into two half-decks and Shuffle back into one full deck
		while ( !L.isEmpty() ) {
			A.insertAfterLast( L.getFirst() ); L.deleteFirst();
			if ( !L.isEmpty() ) {
				B.insertBeforeFirst( L.getLast() ); L.deleteLast();
			}
		}
		while( !A.isEmpty() || !B.isEmpty() ) {
			if ( !A.isEmpty() ) {	L.insertBeforeFirst(A.getLast()); A.deleteLast();	}
			if ( !B.isEmpty() ) {	L.insertBeforeFirst(B.getLast()); B.deleteLast();	}
		}
		// Increment number of shuffles
		shuffles++;
	}
	
	// Other Procedures
	// public static void main(String[] args)
	// runs Shuffle with command-line argument
	public static void main(String[] args) {
		// new deck with number of cards from command line
		new Shuffle( Integer.parseInt(args[0]) );
	}
}

File: pa2/Shuffle.c
/*
	List.c

	Created by Colin Rognlie on 7/3/05.
	Copyright 2005. All rights reserved.

	Colin Rognlie, reflex@cats.ucsc.edu, CMPS101 pa2
*/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

/* Prototypes */
int calcShuffle (int n);
void shuffle(ListRef L);
int main (int argc, char* argv[]);

/* int main (int argc, char* argv[])
// Parses command line input, calls calcShuffle, prints output */
int main (int argc, char* argv[]) {
	int n;
	if(argc != 2) { printf("Usage: %s number\n", argv[0]); exit(1); }
	n = atoi(argv[1]);
	if ( n < 1 ) { printf("Input number must be 1 or greater.\n"); exit(1); }
	printf("%d\n", calcShuffle(n));
	return 0;
}

/* void calcShuffle (int n)
// Calculated number of shuffles to return to starting deck */
int calcShuffle (int n) {
	/* Initialize variables */
	ListRef deck = newList();
	ListRef startdeck = newList();
	int i, shuffles;
	/* Create initial decks */
	for (i=1; i<=n; i++) { insertAfterLast(deck,i); }
	startdeck = copy(deck);
	/* Shuffle until return to starting deck */
	shuffles = 1;
	for( shuffle(deck); !equals(deck,startdeck); shuffle(deck) ) { shuffles++; }
	return shuffles;
}

/* void shuffle(L) // Shuffles deck according to given algorithm */
void shuffle(ListRef L) {
	/* Initialize two half decks */
	ListRef A = newList();
	ListRef B = newList();
	/* Split main deck into top and bottom halves */
	while ( !isEmpty(L) ) {
		insertAfterLast( A,getFirst(L) ); deleteFirst(L);
		if ( !isEmpty(L) ) {
			insertBeforeFirst( B,getLast(L) ); deleteLast(L);
		}
	}
	/* Shuffle halves together back into main deck */
	while ( !isEmpty(A) || !isEmpty(B) ) {
		if ( !isEmpty(A) ) { insertBeforeFirst(L,getLast(A)); deleteLast(A); }
		if ( !isEmpty(B) ) { insertBeforeFirst(L,getLast(B)); deleteLast(B); }
	}
}
